components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -272,133 +314,496 @@ do_registration_ends ()
+  const int num_cols   = m_phys_grid->get_num_local_dofs();
+  const auto lid2elgp  = m_dyn_grid->get_lid_to_idx_map();
+  const auto p2d       = m_p2d;
 
-    const auto& ph = phys.get_header();
-    const auto& dh = dyn.get_header();
+  const auto field_loop = KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const int i = team.league_rank();
 
-    const auto& phys_alloc_prop = ph.get_alloc_properties();
-    const auto& dyn_alloc_prop  = dh.get_alloc_properties();
+    switch (phys_layout(i)) {
+      case 0:
+      case 1:
+      case 2:
+      {
+        if (pack_alloc_property(i) == 0) {
+          set_dyn_to_zero<pack_type>(team);
+          team.team_barrier();
 
-    const bool phys_pack_alloc = phys_alloc_prop.template is_compatible<pack_type>();
-    const bool dyn_pack_alloc  = dyn_alloc_prop.template  is_compatible<pack_type>();
-    const bool phys_small_pack_alloc = phys_alloc_prop.template is_compatible<small_pack_type>();
-    const bool dyn_small_pack_alloc  = dyn_alloc_prop.template  is_compatible<small_pack_type>();
+          local_remap_fwd_2d<pack_type>(team, num_cols, lid2elgp, p2d);
+        } else if (pack_alloc_property(i) == 1) {
+          set_dyn_to_zero<small_pack_type>(team);
+          team.team_barrier();
 
-    const int itl = m_is_state_field[i] ? tl.np1 : -1;
+          local_remap_fwd_2d<small_pack_type>(team, num_cols, lid2elgp, p2d);
+        } else {
+          set_dyn_to_zero<Real>(team);
+          team.team_barrier();
 
-    const LayoutType lt = get_layout_type(ph.get_identifier().get_layout().tags());
-    switch (lt) {
-      case LayoutType::Scalar2D:
-      case LayoutType::Vector2D:
-      case LayoutType::Tensor2D:
-        remap_bwd_2d(phys,dyn,itl);
+          local_remap_fwd_2d<Real>(team, num_cols, lid2elgp, p2d);
+        }
         break;
-      case LayoutType::Scalar3D:
-      case LayoutType::Vector3D:
-      case LayoutType::Tensor3D:
-        if (phys_pack_alloc && dyn_pack_alloc) {
-          remap_bwd_3d_impl<pack_type>(phys,dyn,itl);
-        } else if (phys_small_pack_alloc && dyn_small_pack_alloc) {
-          remap_bwd_3d_impl<small_pack_type>(phys,dyn,itl);
+      }
+      case 3:
+      case 4:
+      case 5:
+      {
+        if (pack_alloc_property(i) == 0) {
+          set_dyn_to_zero<pack_type>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<pack_type>(team, num_cols, lid2elgp, p2d);
+        } else if (pack_alloc_property(i) == 1) {
+          set_dyn_to_zero<small_pack_type>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<small_pack_type>(team, num_cols, lid2elgp, p2d);
         } else {
-          remap_bwd_3d_impl<Real>(phys,dyn,itl);
+          set_dyn_to_zero<Real>(team);
+          team.team_barrier();
+
+          local_remap_fwd_3d<Real>(team, num_cols, lid2elgp, p2d);
         }
         break;
+      }
       default:
         ekat::error::runtime_abort("Error! Unhandled case in switch statement.\n");
```

Comment:
I think `runtime_abort` is not device friendly. You might need a `EKAT_KERNEL_ERROR_MSG` here.